genie-agent/src/main/java/com/netflix/genie/agent/aws/s3/S3ClientFactory.java

```diff
@@ -0,0 +1,241 @@
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenService;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClientBuilder;
+import com.google.common.annotations.VisibleForTesting;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.boot.context.properties.bind.Bindable;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.core.env.Environment;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * An {@link AmazonS3} client factory class. Given {@link AmazonS3URI} instances and the configuration of the system
+ * this factory is expected to return a valid client instance for the S3 URI which can then be used to access that URI.
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@Slf4j
+public class S3ClientFactory {
+
+    @VisibleForTesting
+    static final String BUCKET_PROPERTIES_ROOT_KEY = "genie.aws.s3.buckets";
+
+    private final AWSCredentialsProvider awsCredentialsProvider;
+    private final Map<String, S3ClientKey> bucketToClientKey;
+    private final ConcurrentHashMap<S3ClientKey, AmazonS3> clientCache;
+    private final Map<String, BucketProperties> bucketProperties;
+    private final AWSSecurityTokenService stsClient;
+    private final Regions defaultRegion;
+
+    /**
+     * Constructor.
+     *
+     * @param awsCredentialsProvider The base AWS credentials provider to use for the generated S3 clients
+     * @param regionProvider         How this factory should determine the default {@link Regions}
+     * @param environment            The Spring application {@link Environment}
+     */
+    public S3ClientFactory(
+        final AWSCredentialsProvider awsCredentialsProvider,
+        final AwsRegionProvider regionProvider,
+        final Environment environment
+    ) {
+        this.awsCredentialsProvider = awsCredentialsProvider;
+
+        /*
+         * Use the Spring property binder to dynamically map properties under a common root into a map of key to object.
+         *
+         * In this case we're trying to get bucketName -> BucketProperties
+         *
+         * So if there were properties like:
+         * genie.aws.s3.buckets.someBucket1.roleARN = blah
+         * genie.aws.s3.buckets.someBucket2.region = us-east-1
+         * genie.aws.s3.buckets.someBucket2.roleARN = blah
+         *
+         * The result of this should be two entries in the map "bucket1" and "bucket2" mapping to property binding
+         * object instances of BucketProperties with the correct property set or null if option wasn't specified.
+         */
+        this.bucketProperties = Binder
```

Comment:
It's possibly a reasonable thing to do if we think some of this stuff will be used elsewhere. For the moment i left it in here cause it's all implementation details of dealing with S3 clients and S3 clients only. The `AwsRegionProvider` is passed in and really should never fail as it comes from configuration that already has checks. I really only put all these checks in here in case someone overrode the `AwsRegionProvider` from our auto config with something really bad.

I vote leave it here for now. I think there's some refactoring that needs to be done anyway as we move forward for some of this stuff. e.g. we still need to pass in a `ClientConfiguration` somehow that lets us configure retries and whatnot for the S3 clients. I'm not sure if that's something that we should do directly or PR into spring cloud aws.